File: solutions/1712. Ways to Split Array Into Three Subarrays/1712-2.py
class Solution:
  def waysToSplit(self, nums: List[int]) -> int:
    kMod = int(1e9 + 7)
    n = len(nums)
    ans = 0
    prefix = list(itertools.accumulate(nums))

    j = 0
    k = 0
    for i in range(n - 2):
      # find first index j s.t.
      # left = prefix[i] <= mid = prefix[j] - prefix[i]
      while j <= i or (j < n - 1 and prefix[i] > prefix[j] - prefix[i]):
        j += 1
      # find first index k s.t.
      # mid = prefix[k] - prefix[i] > right = prefix[-1] - prefix[k]
      while k < j or (k < n - 1 and prefix[k] - prefix[i] <= prefix[-1] - prefix[k]):
        k += 1
      ans = (ans + k - j) % kMod

    return ans

File: solutions/1712. Ways to Split Array Into Three Subarrays/1712-2.java
class Solution {
  public int waysToSplit(int[] nums) {
    final int kMod = (int) 1e9 + 7;
    final int n = nums.length;
    int ans = 0;
    int[] prefix = nums.clone();

    for (int i = 1; i < n; ++i)
      prefix[i] += prefix[i - 1];

    for (int i = 0, j = 0, k = 0; i < n - 2; ++i) {
      // find first index j s.t.
      // left = prefix[i] <= mid = prefix[j] - prefix[i]
      while (j <= i || (j < n - 1 && prefix[i] > prefix[j] - prefix[i]))
        ++j;
      // find first index k s.t.
      // mid = prefix[k] - prefix[i] > right = prefix[-1] - prefix[k]
      while (k < j || (k < n - 1 && prefix[k] - prefix[i] <= prefix[prefix.length - 1] - prefix[k]))
        ++k;
      ans = (ans + k - j) % kMod;
    }

    return ans;
  }
}
